consensus/driver/driver_test.go

```diff
@@ -0,0 +1,223 @@
+}
+
+func (m *mockListener[M]) Listen() <-chan M {
+	return m.ch
+}
+
+type mockBroadcaster[M tendermint.Message[value, felt.Felt, felt.Felt]] struct {
+	wg                  sync.WaitGroup
+	mu                  sync.Mutex
+	broadcastedMessages []M
+}
+
+func (m *mockBroadcaster[M]) Broadcast(msg M) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.broadcastedMessages = append(m.broadcastedMessages, msg)
+	m.wg.Done()
+}
+
+func (m *mockBroadcaster[M]) SendMsg(addr felt.Felt, msg M) {
+}
+
+func mockListeners(
+	proposalCh chan tendermint.Proposal[value, felt.Felt, felt.Felt],
+	prevoteCh chan tendermint.Prevote[felt.Felt, felt.Felt],
+	precommitCh chan tendermint.Precommit[felt.Felt, felt.Felt],
+) driver.Listeners[value, felt.Felt, felt.Felt] {
+	return driver.Listeners[value, felt.Felt, felt.Felt]{
+		ProposalListener:  newMockListener(proposalCh),
+		PrevoteListener:   newMockListener(prevoteCh),
+		PrecommitListener: newMockListener(precommitCh),
+	}
+}
+
+func mockBroadcasters() driver.Broadcasters[value, felt.Felt, felt.Felt] {
+	return driver.Broadcasters[value, felt.Felt, felt.Felt]{
+		ProposalBroadcaster:  &mockBroadcaster[tendermint.Proposal[value, felt.Felt, felt.Felt]]{},
+		PrevoteBroadcaster:   &mockBroadcaster[tendermint.Prevote[felt.Felt, felt.Felt]]{},
+		PrecommitBroadcaster: &mockBroadcaster[tendermint.Precommit[felt.Felt, felt.Felt]]{},
+	}
+}
+
+func mockTimeoutFn(step tendermint.Step, round tendermint.Round) time.Duration {
+	return 1 * time.Millisecond
+}
+
+func getRandMessageHeader() tendermint.MessageHeader[felt.Felt] {
+	return tendermint.MessageHeader[felt.Felt]{
+		Height: tendermint.Height(rand.Uint32()),
+		Round:  tendermint.Round(rand.Int()),
+		Sender: felt.FromUint64(rand.Uint64()),
+	}
+}
+
+func getRandProposal() tendermint.Proposal[value, felt.Felt, felt.Felt] {
+	return tendermint.Proposal[value, felt.Felt, felt.Felt]{
+		MessageHeader: getRandMessageHeader(),
+		Value:         utils.HeapPtr(value(rand.Uint64())),
+		ValidRound:    tendermint.Round(rand.Int()),
+	}
+}
+
+func getRandPrevote() tendermint.Prevote[felt.Felt, felt.Felt] {
+	return tendermint.Prevote[felt.Felt, felt.Felt]{
+		MessageHeader: getRandMessageHeader(),
+		ID:            utils.HeapPtr(felt.FromUint64(rand.Uint64())),
+	}
+}
+
+func getRandPrecommit() tendermint.Precommit[felt.Felt, felt.Felt] {
+	return tendermint.Precommit[felt.Felt, felt.Felt]{
+		MessageHeader: getRandMessageHeader(),
+		ID:            utils.HeapPtr(felt.FromUint64(rand.Uint64())),
+	}
+}
+
+func getRandTimeout(step tendermint.Step) tendermint.Timeout {
+	return tendermint.Timeout{
+		Height: tendermint.Height(rand.Uint32()),
+		Step:   step,
+		Round:  tendermint.Round(rand.Int()),
+	}
+}
+
+const (
+	actionCount = 10
+)
+
+func registerActions(expectedBroadcast *expectedBroadcast) []tendermint.Action[value, felt.Felt, felt.Felt] {
+	actions := make([]tendermint.Action[value, felt.Felt, felt.Felt], actionCount)
+	for i := range actionCount {
+		switch rand.Int() % 3 {
```

Comment:
Should we use a seed for deterministic randomization?

Reply:
Agree, I'll update the code